presto-main/src/main/java/com/facebook/presto/operator/ArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,331 @@
+
+        int[] positions = getPositions();
+        verify(positionsOffset + batchSize - 1 < positions.length && positions[positionsOffset] >= 0 &&
+                positions[positionsOffset + batchSize - 1] < columnarArray.getPositionCount());
+
+        appendNulls();
+        appendOffsets();
+        rawBlockBuffer.copyValues();
+
+        bufferedPositionCount += batchSize;
+    }
+
+    @Override
+    public void serializeTo(SliceOutput sliceOutput)
+    {
+        requireNonNull(sliceOutput, "sliceOutput is null");
+
+        // TODO: getSerializedSizeInBytes() was calculated in flush and doesnt need to recalculated
+        verify(getSerializedSizeInBytes() <= sliceOutput.writableBytes());
+
+        writeLengthPrefixedString(sliceOutput, NAME);
+
+        rawBlockBuffer.serializeTo(sliceOutput);
+
+        sliceOutput.writeInt(bufferedPositionCount); //positionCount
+
+        sliceOutput.writeInt(0);  // the base position
+        sliceOutput.appendBytes(offsetsBuffer, 0, offsetsBufferIndex);
+
+        serializeNullsTo(sliceOutput);
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return getPositionsSizeInBytes() + // positions and mappedPositions
+                offsetsBufferIndex +
+                getNullsBufferSizeInBytes();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPostionsRetainedSizeInBytes() +
+                sizeOf(offsetsBuffer) +
+                getNullsBufferRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +            // encoding name
+                rawBlockBuffer.getSerializedSizeInBytes() +   // nested block
+                SIZE_OF_INT +                           // positionCount
+                SIZE_OF_INT +                           // offset 0. The offsetsBuffer doesn't contain the offset 0 so we need to add it here.
+                offsetsBufferIndex +                    // offsets buffer.
+                getNullsBufferSerializedSizeInBytes();  // nulls
+    }
+
+    private void mapPositions(DecodedBlockNode decodedBlockNode)
+    {
+        // Remap positions
+        Object decodedObject = decodedBlockNode.getDecodedBlock();
+
+        if (decodedObject instanceof DictionaryBlock) {
+            DictionaryBlock dictionaryBlock = (DictionaryBlock) decodedObject;
+            mapPositionsForDictionary(dictionaryBlock);
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            isPositionsMapped = true;
+        }
+        else if (decodedObject instanceof RunLengthEncodedBlock) {
+            mapPositionsForRle();
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            isPositionsMapped = true;
+        }
+        else {
+            isPositionsMapped = false;
+        }
+
+        columnarArray = (ColumnarArray) decodedBlockNode.getDecodedBlock();
```

Comment:
This logic doesn't seem to belong here.  Consider moving it to the caller.

```
    @Override
    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
    {
        requireNonNull(decodedBlockNode, "decodedBlockNode is null");

        columnarArray = (ColumnarArray) mapPositions(decodedBlockNode).getDecodedBlock();
        decodedBlock = columnarArray.getNullCheckBlock();

        populateNestedPositions();
        elementsBuffers.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
    }

    private DecodedBlockNode mapPositions(DecodedBlockNode decodedBlockNode)
    {
        Object decodedObject = decodedBlockNode.getDecodedBlock();

        if (decodedObject instanceof DictionaryBlock) {
            mapPositionsForDictionary((DictionaryBlock) decodedObject);
            isPositionsMapped = true;
            return decodedBlockNode.getChildren().get(0);
        }

        if (decodedObject instanceof RunLengthEncodedBlock) {
            mapPositionsForRle();
            isPositionsMapped = true;
            return decodedBlockNode.getChildren().get(0);
        }

        isPositionsMapped = false;
        return decodedBlockNode;
    }
```

This makes `mapPositions` method the same for all complex types and it feels like it should apply to primitive types as well. It should be possible to move this method into `BlockEncodingBuffers` and re-write `BlockEncodingBuffers#setupDecodedBlockAndMapPositions` as follows:

```
    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
    {
        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
        decodedBlock = (Block) mapPositions(decodedBlockNode).getDecodedBlock();
    }
```